File: 11_Container_With_Most_Water/two_pointer.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int max_capacity = 0;
        int capacity;
        int i = 0;
        int j = height.size() - 1;
        
        while (i != j){
            if (height[i] < height[j]){
                capacity = height[i] * (j - i);
                i++;
            }else{
                capacity = height[j] * (j - i);
                j--;
            }
            if (capacity > max_capacity){
                max_capacity = capacity;
            }

        }
        return max_capacity;
    }
};


// Time Complexity: O(n)
// Runtime: 99 ms	Memory: 59 MB

File: 11_Container_With_Most_Water/two_pointer.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        max_capacity = 0
        i = 0
        j = len(height) - 1
        
        for _ in range(len(height) - 1):
            if height[i] < height[j]:
                capacity = height[i] * (j - i)
                i += 1
            else:
                capacity = height[j] * (j - i)
                j -= 1
                
            if capacity > max_capacity:
                max_capacity = capacity
                
        return max_capacity


# Time Complexity: O(n)
# Runtime: 782 ms	Memory: 27.5 MB